rtrlib/rtr/packets.c

```diff
@@ -192,102 +209,136 @@ void rtr_change_socket_state(struct rtr_socket *rtr_socket, const enum rtr_socke
         rtr_socket->connection_state_fp(rtr_socket, new_state,rtr_socket->connection_state_fp_param);
 }
 
-static void rtr_pdu_to_network_byte_order(void *pdu)
+static void rtr_pdu_convert_header_byte_order(void *pdu, enum target_byte_order target_byte_order)
 {
-    struct pdu_header *header = pdu;
-    struct pdu_error *err_pdu = NULL;
+	uint16_t (*convert_short)(uint16_t);
+	uint32_t (*convert_long)(uint32_t);
+
+	if (target_byte_order == TO_NETWORK_BYTE_ORDER) {
+		convert_short = &htons;
+		convert_long = &htonl;
+	} else if (target_byte_order == TO_HOST_HOST_BYTE_ORDER) {
+		convert_short = &ntohs;
+		convert_long = &ntohl;
+	} else {
+		assert(0);
+	}
 
-    header->reserved = htons(header->reserved);
-    header->len = htonl(header->len);
+	struct pdu_header *header = pdu;
 
-    const enum pdu_type type = rtr_get_pdu_type(pdu);
-    switch (type) {
-    case SERIAL_QUERY:
-        ((struct pdu_serial_query *) pdu)->sn = htonl(((struct pdu_serial_query *) pdu)->sn);
-        break;
-    case ERROR:
-        err_pdu = pdu;
-        *((uint32_t *)(err_pdu->rest + err_pdu->len_enc_pdu)) =
-            htonl(*((uint32_t *)(err_pdu->rest + err_pdu->len_enc_pdu)));
-        err_pdu->len_enc_pdu = htonl(err_pdu->len_enc_pdu);
-        break;
-    default:
-        break;
-    }
+	//The ROUTER_KEY PDU has two 1 Byte fields instead of the 2 Byte reserved field.
+	if (header->type != ROUTER_KEY) {
+		header->reserved = convert_short(header->reserved);
+	}
+
+	header->len = convert_long(header->len);
 }
 
-static void rtr_pdu_footer_to_host_byte_order(void *pdu)
-{
-    const enum pdu_type type = rtr_get_pdu_type(pdu);
-    struct pdu_header *header = pdu;
-    struct pdu_error * err_pdu = NULL;
+static void rtr_pdu_convert_footer_byte_order(void *pdu, enum target_byte_order target_byte_order) {
+	uint32_t (*convert_long)(uint32_t);
+	struct pdu_error *err_pdu;
+	struct pdu_header *header = pdu;
+	uint32_t addr6[4];
+	const enum pdu_type type = rtr_get_pdu_type(pdu);
+
+	if (target_byte_order == TO_NETWORK_BYTE_ORDER) {
```

Comment:
This code is duplicate with line 217 in function `rtr_pdu_convert_header_byte_order()`. Consider `get_convert_short_fn` and `get_convert_long_fn` as static functions with parameter `enum target_byte_order` and returning pointer to functions `htons`/`ntohs` or `htonl`/`ntohl`